File: ai/bot/helper/collect_item.py
##
## EPITECH PROJECT, 2025
## B-YEP-400-STG-4-1-zappy-noe.carabin
## File description:
## collect_item
##

from . import find_item as fi
from ...network import ai_socket as aso

def collect_item(look_string, item_name, sock, reader, food_level, inventory, level,
    recipes, thread_name = "UnknownThread"):
    """Take an item if founded on current tile and not too present in inventory."""

    if fi.find_item_in_look(look_string, item_name, thread_name) == 0:
        if ((item_name == 'food' and food_level >= 12.0) or level == 8 or (item_name != 'food' and
        inventory.get(item_name, 0) >= recipes[level - 1][item_name] * 2 + 1)):
            return True
        return aso.do_action(sock, reader, f"Take {item_name}", thread_name) != "dead"
    return True

def collect_all_items(look_string, sock, reader, food_level, inventory, level, recipes, thread_name = "UnknownThread"):
    """Try to take all the items on the current tile."""

    return (collect_item(look_string, 'food', sock, reader, food_level, inventory, level, recipes, thread_name) and
    collect_item(look_string, 'linemate', sock, reader, food_level, inventory, level, recipes, thread_name) and
    collect_item(look_string, 'deraumere', sock, reader, food_level, inventory, level, recipes, thread_name) and
    collect_item(look_string, 'sibur', sock, reader, food_level, inventory, level, recipes, thread_name) and
    collect_item(look_string, 'mendiane', sock, reader, food_level, inventory, level, recipes, thread_name) and
    collect_item(look_string, 'phiras', sock, reader, food_level, inventory, level, recipes, thread_name) and
    collect_item(look_string, 'thystame', sock, reader, food_level, inventory, level, recipes, thread_name))

File: ai/entity/bot/helper/collect_item.cpp
/*
** EPITECH PROJECT, 2025
** B-YEP-400-STG-4-1-zappy-noe.carabin
** File description:
** collect_item
*/

/**
 * @file collect_item.cpp
 * @brief Implementation of item collection functions.
 * @date 29/06/2025
 * @author Jason KOENIG
 *
 * @see AI.hpp
 */

#include "../../AI.hpp"
#include "../../../utils/utils.hpp"

bool ai::entity::AI::collectItem(const std::string &look, const std::string &name,
    int recipe_qty, bool is_players)
{
    int pos = findItemInLook(look, name);

    if (pos == -1)
        return false;
    if (pos != 0)
        return true;
    if ((name == "food" && _food_level >= MAX_FOOD_THRESHOLD) || _level == 8 ||
    (name != "food" && _inventory.getItem(name) >= recipe_qty * 2 + 1) || is_players)
        return true;
    return doKoAction("Take " + name);
}

bool ai::entity::AI::collectAllItems(const std::string &look)
{
    const bool is_player = countItemOnTile(0, look, "player") > 0;

    return collectItem(look, "food", _food_level, is_player) &&
    collectItem(look, "linemate", RECIPES[_level - 1].linemate, is_player) &&
    collectItem(look, "deraumere", RECIPES[_level - 1].deraumere, is_player) &&
    collectItem(look, "sibur", RECIPES[_level - 1].sibur, is_player) &&
    collectItem(look, "mendiane", RECIPES[_level - 1].mendiane, is_player) &&
    collectItem(look, "phiras", RECIPES[_level - 1].phiras, is_player) &&
    collectItem(look, "thystame", RECIPES[_level - 1].thystame, is_player);
}
